File: tests/smith-waterman/cpp/test_smith_waterman.cpp
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <tuple>

#include "impl.hpp"          // staged adapter: class SmithWaterman with findAlignment(query, reference)
#include "bench_utils.hpp"

using namespace std;

static int g_failures = 0;
static const double EPS = 1e-6;

static void check(bool ok, const string& name) {
    cout << name << (ok ? " passed" : " failed") << endl;
    if (!ok) g_failures++;
}

static void test_case(const string& name, const string& query, const string& reference,
                      const string& expA, const string& expB, int expScore, double expIdentity) {
    SmithWaterman sw(2, -1, -1);
    auto [a, b, score, identity] = sw.findAlignment(query, reference);
    bool ok = (a == expA) && (b == expB) && (score == expScore)
              && (fabs(identity - expIdentity) <= EPS);
    if (!ok) {
        cout << "  got: a=" << a << " b=" << b << " score=" << score
             << " identity=" << identity << endl;
    }
    check(ok, name);
}

static bool readSWInput(string& query, string& reference, int& score, double& identity) {
    const char* path = getenv("SW_INPUT");
    if (!path) return false;
    ifstream f(path);
    if (!f) return false;
    string sScore, sIdentity;
    getline(f, query);
    getline(f, reference);
    getline(f, sScore);
    getline(f, sIdentity);
    score = stoi(sScore);
    identity = stod(sIdentity);
    return true;
}

int main() {
    test_case("test_perfect_match", "ACGT", "ACGT", "ACGT", "ACGT", 8, 100.0);
    test_case("test_no_match", "AAAA", "TTTT", "", "", 0, 0.0);
    test_case("test_partial_match", "ACGTACGT", "TACGTGCA", "TACGT", "TACGT", 10, 100.0);
    test_case("test_gap_in_query", "ACGT", "ACAGT", "AC-GT", "ACAGT", 7, 80.0);
    test_case("test_gap_in_reference", "ACAGT", "ACGT", "ACAGT", "AC-GT", 7, 80.0);
    test_case("test_single_char_match", "A", "A", "A", "A", 2, 100.0);

    string query, reference;
    int expScore = 0;
    double expIdentity = 0.0;
    bool haveInput = readSWInput(query, reference, expScore, expIdentity);

    if (haveInput) {
        SmithWaterman sw(2, -1, -1);
        auto [a, b, score, identity] = sw.findAlignment(query, reference);
        bool ok = (a.size() == b.size()) && (score == expScore)
                  && (fabs(identity - expIdentity) <= EPS);
        check(ok, "test_large_sequences_from_file");
    }

    if (g_failures > 0) {
        cout << g_failures << " test(s) failed — skipping benchmark" << endl;
        return 1;
    }

    if (!haveInput) {
        cout << "SW_INPUT not set — skipping benchmark" << endl;
        return 0;
    }

    SmithWaterman sw(2, -1, -1);
    int reps = bench_reps(5), iters = bench_iters(1);
    auto bm = run_benchmark([&]() { sw.findAlignment(query, reference); }, reps, iters, 1);
    cout << format_result("SW large sequences", bm) << endl;

    string params = "{\"query_len\": " + to_string(query.size())
        + ", \"reference_len\": " + to_string(reference.size()) + "}";
    write_result(bm, bench_out(), "smith-waterman", "cpp", bench_impl(), iters, params);
    return 0;
}

File: tests/smith-waterman/python/test_smith_waterman.py
import os
import unittest

import pytest

from tests import bench_utils
# The implementation under test is staged by bench/run.py (see
# bench/manifests/smith-waterman/python.json); the adapter exposes the
# canonical entry (a SmithWaterman class with .findAlignment).
from .staging.impl_adapter import BenchSmithWaterman as SmithWaterman

CONFIG = bench_utils.get_bench_config(default_reps=5, default_iters=1)

# Shared large-sequence input, path provided by the runner (SW_INPUT env).
SW_INPUT = os.environ.get(
    "SW_INPUT",
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "data", "large_test_input.txt"),
)


class TestSmithWaterman(unittest.TestCase):
    """Correctness via the shared findAlignment entry (uniform across all impls)."""

    def setUp(self):
        self.sw = SmithWaterman(matchScore=2, mismatchScore=-1, gapScore=-1)
        self.e = 1e-6

    def test_perfect_match(self):
        a, b, score, identity = self.sw.findAlignment("ACGT", "ACGT")
        self.assertEqual(a, b)
        self.assertEqual(score, 8)
        self.assertAlmostEqual(identity, 100.0, delta=self.e)

    def test_no_match(self):
        a, b, score, identity = self.sw.findAlignment("AAAA", "TTTT")
        self.assertEqual(a, "")
        self.assertEqual(b, "")
        self.assertEqual(score, 0)
        self.assertAlmostEqual(identity, 0.0, delta=self.e)

    def test_partial_match(self):
        a, b, score, identity = self.sw.findAlignment("ACGTACGT", "TACGTGCA")
        self.assertEqual(a, "TACGT")
        self.assertEqual(b, "TACGT")
        self.assertEqual(score, 10)
        self.assertAlmostEqual(identity, 100.0, delta=self.e)

    def test_gap_in_query(self):
        a, b, score, identity = self.sw.findAlignment("ACGT", "ACAGT")
        self.assertEqual(a, "AC-GT")
        self.assertEqual(b, "ACAGT")
        self.assertEqual(score, 7)
        self.assertAlmostEqual(identity, 80.0, delta=self.e)

    def test_gap_in_reference(self):
        a, b, score, identity = self.sw.findAlignment("ACAGT", "ACGT")
        self.assertEqual(a, "ACAGT")
        self.assertEqual(b, "AC-GT")
        self.assertEqual(score, 7)
        self.assertAlmostEqual(identity, 80.0, delta=self.e)

    def test_single_character_match(self):
        a, b, score, identity = self.sw.findAlignment("A", "A")
        self.assertEqual(a, "A")
        self.assertEqual(b, "A")
        self.assertEqual(score, 2)
        self.assertAlmostEqual(identity, 100.0, delta=self.e)

    def test_single_character_mismatch(self):
        a, b, score, identity = self.sw.findAlignment("A", "T")
        self.assertEqual(score, 0)
        self.assertAlmostEqual(identity, 0.0, delta=self.e)

    def test_large_sequences_from_file(self):
        with open(SW_INPUT) as f:
            lines = f.readlines()
        query, reference = lines[0].strip(), lines[1].strip()
        expected_score = int(lines[2].strip())
        expected_identity = float(lines[3].strip())

        a, b, score, identity = self.sw.findAlignment(query, reference)
        self.assertEqual(len(a), len(b))
        self.assertEqual(score, expected_score)
        self.assertAlmostEqual(identity, expected_identity, delta=self.e)


class TestSmithWatermanPerformance(unittest.TestCase):

    @pytest.mark.perf
    def test_performance_large_sequences(self):
        with open(SW_INPUT) as f:
            lines = f.readlines()
        query, reference = lines[0].strip(), lines[1].strip()

        sw = SmithWaterman(matchScore=2, mismatchScore=-1, gapScore=-1)
        reps, iters = CONFIG["reps"], CONFIG["iters"]
        result = bench_utils.run_benchmark(
            lambda: sw.findAlignment(query, reference), reps=reps, iters=iters, warmup=1
        )
        print(bench_utils.format_result("SW large sequences", result))
        bench_utils.write_result(
            result, algo="smith-waterman", lang="python", impl=CONFIG["impl"],
            iters_per_rep=iters,
            params={"query_len": len(query), "reference_len": len(reference)},
        )
